numpy/core/src/umath/matmul.c.src

```diff
@@ -0,0 +1,393 @@
+/* -*- c -*- */
+
+#define _UMATHMODULE
+#define _MULTIARRAYMODULE
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+
+#include "Python.h"
+
+#include "npy_config.h"
+#include "numpy/npy_common.h"
+#include "numpy/arrayobject.h"
+#include "numpy/ufuncobject.h"
+#include "numpy/npy_math.h"
+#include "numpy/halffloat.h"
+#include "lowlevel_strided_loops.h"
+
+#include "npy_pycompat.h"
+
+#include "npy_cblas.h"
+#include "arraytypes.h" /* For TYPE_dot functions */
+#include <assert.h>
+
+/*
+ *****************************************************************************
+ **                            BASICS                                       **
+ *****************************************************************************
+ */
+
+/*
+ * Determine if a non-contiguous axis has strides and length supported by blas.
+ * 1. Strides must be multiples of the itemsize, and must not cause segments of
+ *    the contiguous axis to overlap.
+ * 2. BLAS requires sizes to fit within `int`s.
+ */
+
+#define BLAS_ABLE(n, stride, itemsize) ((stride % itemsize == 0) && \
+                                  (stride / itemsize >= n) && \
+                                  (n < NPY_MAX_INT) && \
+                                  (stride / itemsize < NPY_MAX_INT))
+
+#if defined(HAVE_CBLAS)
+static const npy_cdouble oneD = {1.0, 0.0}, zeroD = {0.0, 0.0};
+static const npy_cfloat  oneF = {1.0, 0.0}, zeroF = {0.0, 0.0};
+
+/**begin repeat
+ *
+ * #name = FLOAT, DOUBLE, CFLOAT, CDOUBLE#
+ * #ctype = npy_float, npy_double, npy_cfloat, npy_cdouble#
+ * #typ = npy_float, npy_double, npy_cfloat, npy_cdouble#
+ * #prefix = s, d, c, z#
+ * #step1 = 1.F, 1., &oneF, &oneD#
+ * #step0 = 0.F, 0., &zeroF, &zeroD#
+ */
+NPY_NO_EXPORT void
+@name@_gemv(void *ip1, npy_intp is1_m, npy_intp is1_n, void *ip2, npy_intp is2_n, void *op,
+               npy_intp m, npy_intp n)
+{
+    /*
+     * Vector matrix multiplication -- Level 2 BLAS
+     * arguments
+     * ip1: contiguous data, m*n shape
+     * ip2: data in c order, n*1 shape
+     * op:  contiguous data in c order, m shape
+     */
+    enum CBLAS_ORDER order;
+    int M, N, lda;
+
+    assert (BLAS_ABLE(n, is1_m, sizeof(@typ@)));
+    assert (BLAS_ABLE(m, is1_n, sizeof(@typ@)));
+    assert (BLAS_ABLE(m, is2_n, sizeof(@typ@)));
```

Comment:
Are you sure these `n`s and `m`s aren't swapped?

Reply:
the assert was failing in the debug CI test, fixing.

Reply:
I'm wrong, these `m`s and `n`s are correct.

What's wrong is that the asserts on `i1` should be within the `if` or `if-else`

Reply:
I rewrote the macro instead